flink-runtime/src/test/java/org/apache/flink/runtime/metrics/util/MetricUtilsTest.java

```diff
@@ -243,4 +247,84 @@ public MetricGroup addGroup(String name) {
                                 METRIC_GROUP_MEMORY,
                                 METRIC_GROUP_MANAGED_MEMORY)));
     }
+
+    // --------------- utility methods and classes ---------------
+
+    /** Intercept loading classes and define them based on found class resources. */
+    private static class InterceptingClassLoader extends ClassLoader {
+        private final Set<String> interceptingClassNames = new HashSet<>();
+        private final Map<String, Class<?>> interceptedClasses = new HashMap<>();
+
+        public InterceptingClassLoader(
+                Collection<String> interceptingClassNames, ClassLoader parent) {
+            super(parent);
+            this.interceptingClassNames.addAll(interceptingClassNames);
+        }
+
+        @Override
+        protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+            Class<?> interceptedClass = interceptedClasses.get(name);
+            if (interceptedClass != null) {
+                return interceptedClass;
+            }
+            if (!interceptingClassNames.contains(name)) {
+                return super.loadClass(name, resolve);
+            }
+            String classFileResourceName = name.replace('.', '/') + ".class";
+            try (InputStream inputStream = getResourceAsStream(classFileResourceName)) {
+                ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+                IOUtils.copyBytes(inputStream, outputStream, false);
+                byte[] bytecode = outputStream.toByteArray();
+                Class<?> definedClass =
+                        defineClass(
+                                name,
+                                bytecode,
+                                0,
+                                bytecode.length,
+                                getClass().getProtectionDomain());
+                interceptedClasses.put(name, definedClass);
+                return definedClass;
+            } catch (Exception ex) {
+                throw new ClassNotFoundException(ex.getMessage(), ex);
+            }
+        }
+    }
+
+    /** Define an new class using given class's name and bytecode. */
+    @SuppressWarnings("unchecked")
+    private static <T> Class<T> redefineAsNewClass(Class<T> clazz) throws ClassNotFoundException {
+        InterceptingClassLoader classLoader =
+                new InterceptingClassLoader(
+                        Collections.singleton(clazz.getName()), clazz.getClassLoader());
+        Class<T> newClass = (Class<T>) classLoader.loadClass(clazz.getName(), true);
```

Comment:
```suggestion
    private static Class<?> redefineAsNewClass(Class<?> clazz) throws ClassNotFoundException {
        ClassLoader classLoader =
                new ChildFirstClassLoader(
                        ClassLoaderUtils.getClasspathURLs(),
                        clazz.getClassLoader(),
                        new String[] {},
                        null);
        Class<?> newClass = classLoader.loadClass(clazz.getName());
```
This seems to be quite stable without requiring this whole InterceptingClassLoader thing.

Reply:
Hi @zentol , this does not work in my IDEA with `java.lang.SecurityException: Prohibited package name: java.lang`, but does work in `mvn test`.

The cause is that IDEA count `rt.jar` to `java.class.path`. Then it will try to load `java.lang.Object`.

Reply:
I can see why that would happen; for that reason all java classes are included in `classloader.parent-first-patterns.default`. I do wonder though why this works fine on ma machine.

In any case, we should be able to remedy that by loading a class that has no further dependencies (like `private static class Dummy {}`) and/or passing `new String[] {"java."}` as `alwaysParentFirstPatterns`.

Reply:
I think we should resort to pass additional parent first pattern, `Dummy` call requires `java.lang.Object` anyway.

Reply:
@zentol I have pushed new fixup commit to go through the "and" path with `ChilldFirstClassLoader`. Combinations of "or" paths do not work well/stable with `ChildFirstClassLoader`. Please take another look.